### PR TITLE
SOF-1747/active rundown prompt

### PR DESCRIPTION
**Pull request by k-primeholding-dim**
With this PR, oc Rundown activation a confirmation dialog will open that will warn if currently active rundown will be deactivated before activating the one that we intend to.

In order to get the info - a broad getBasicRundowns is introduced to BasicRundownStatService which results are filtered by mode.
Also a new DialogConfirmationService is added to accommodate switchActivateRundownDialog() and switchRehearsalRundownDialog() that are used for user interaction when manipulating a Rundown to On-Air and to Rehearsal mode.

### CODE REVIEW

**Review by RasmusAlbrektsen — COMMENTED** [review]
I think it would be smart to include the rehearsal changes from develop and add this functionality for `REHEARSAL` mode since this is needed as well. I can see it is not explicitly written in the task but it was, however, blocked by those changes so i guess it was the intention.

**Comment by LindvedKrvang** [comment]
> I think it would be smart to include the rehearsal changes from develop and add this functionality for `REHEARSAL` mode since this is needed as well. I can see it is not explicitly written in the task but it was, however, blocked by those changes so i guess it was the intention.

Rehearsal it's part of the acceptance criteria, så the task isn't done before it also works for rehearsal
image

File: src/app/rundown-view/factories/key-binding.factory.ts
Comment (LindvedKrvang) [inline src/app/rundown-view/factories/key-binding.factory.ts:158]: This is duplicated with the code in the RundownHeaderContextMenu.
Whenever you start writing duplicated code, something in you should trigger and you should try to see if it's possible to avoid it. In  this case, there is not really anything that is component specific, so the code could be extracted to a service. Just to give an example.

**Review by LindvedKrvang — CHANGES_REQUESTED** [review]
I haven't reviewed the code, but testing the changes reveals some bugs:
- The user is no longer being asked to confirm if he wants to activate a Rundown, when there are no other Rundowns activated or in Rehearsal.
- There is no confirmation box _whatsoever_ when trying to put a Rundown in Rehearsal. If there is any other Rundown that is Active or in Rehearsal, they are just deactivated with no warning or notification...

**Review by LindvedKrvang — COMMENTED** [review]
Overall it looks good :)

Main thing is that you have given `DialogService` a responsibility that it shouldn't have.

Also, a nice to have, would be to make the Confirmation Dialog red when another Rundown needs to be deactivated. Right now, it's quite easy to miss because the only thing that indicates is a text string. So having something visual different would be a good thing.
All you need to do is to provide `DialogServerity.DANGER` to the confirm dialog when another Rundown would be deactivated.

File: src/locale/messages.da.xlf
Comment (k-primeholding-dim) [reply src/locale/messages.da.xlf]: You are correct - `\n` does not break the line, but simple multi-line translation resource string can be represented as is written in the resource file (with new line) if we use 
```
.mat-dialog-content {
    white-space: pre-line;
}
```
which is what I did.

File: src/app/rundown-view/factories/key-binding.factory.ts
Comment (KvelaGorrrrnio) [inline src/app/rundown-view/factories/key-binding.factory.ts]: The complexity of this method is slowly increasing due to the overwrites of values and conditional updates. I suggest that the creation of key bindings are split up into groups which are based on the rundown state. My suggestion is to have the creation of rundown mode transitions key bindings (activation, entering rehearsal and deactivation) and the rundown interaction key bindings (reset, take, set next, mini shelf cycle) into their own categories and then use composition to chain them together. I have provided a suggestion below with this categorization.

Sidenote: It also looks like that entering rehearsal have no key binding yet, so I have added it in the suggestion with the key binding `Control + Backquote`.

```typescript
class KeyBindingFactory {
  public createRundownKeyBindings(rundown: Rundown): StyledKeyBinding[] {
    return [
      ...this.createRundownModeTransitionKeyBindings(rundown),
      ...this.createRundownInteractionKeyBindings(rundown),
    ]
  }

  private createRundownModeTransitionKeyBindings(rundown: Rundown): StyledKeyBinding[] {
    switch (rundown.mode) {
      case RundownMode.ACTIVE:
        return [
          this.createRundownKeyBinding('Deactivate Rundown', ['Control', 'Shift', 'Backquote'], () => this.deactivateRundown(rundown)),
        ]
      case RundownMode.REHEARSAL:
        return [
          this.createRundownKeyBinding('Activate Rundown', ['Backquote'], () => this.dialogConfirmationService.switchActivateRundownDialog(rundown)),
          this.createRundownKeyBinding('Deactivate Rundown', ['Control', 'Shift', 'Backquote'], () => this.deactivateRundown(rundown)),
        ]
      case RundownMode.INACTIVE:
        return [
          this.createRundownKeyBinding('Activate Rundown', ['Backquote'], () => this.dialogConfirmationService.switchActivateRundownDialog(rundown)),
          this.createRundownKeyBinding('Enter Rehearsal', ['Control', 'Backquote'], () => this.dialogConfirmationService.switchRehearsalRundownDialog(rundown)),
        ]
    }
  }

  private createRundownInteractionKeyBindings(rundown: Rundown): StyledKeyBinding[] {
    switch (rundown.mode) {
      case RundownMode.ACTIVE:
      case RundownMode.REHEARSAL:
        return [
          this.createRundownKeyBinding('Take', ['AnyEnter'], () => this.takeNext(rundown)),
          this.createRundownKeyBinding('Reset Rundown', ['Escape'], () => this.resetRundown(rundown)),
          this.createRundownKeyBinding('Set Segment Above as Next', ['Shift', 'ArrowUp'], () => this.setSegmentAboveNextAsNext(rundown)),
          this.createRundownKeyBinding('Set Segment Below as Next', ['Shift', 'ArrowDown'], () => this.setSegmentBelowNextAsNext(rundown)),
          this.createRundownKeyBinding('Set Earlier Part as Next', ['Shift', 'ArrowLeft'], () => this.setEarlierPartAsNext(rundown)),
          this.createRundownKeyBinding('Set Later Part as Next', ['Shift', 'ArrowRight'], () => this.setLaterPartAsNext(rundown)),
          this.createRundownKeyBinding('Cycle MiniShelf', ['Tab'], () => this.miniShelfCycleService.cycleMiniShelfForward(rundown)),
          this.createRundownKeyBinding('Cycle MiniShelf', ['Shift', 'Tab'], () => this.miniShelfCycleService.cycleMiniShelfBackward(rundown)),
        ]
      case RundownMode.INACTIVE:
        return [
          this.createRundownKeyBinding('Reset Rundown', ['Escape'], () => this.resetRundown(rundown)),
        ]
    }
  }
}
```
In the suggestion I have used a switch statement to differentiate between what system actions are available for the given rundown modes. It makes it easy to extend and have a low complexity.

It should be noted that this file is deleted in #142. Just so we are aware of it.

File: src/app/shared/services/dialog-confirmation.service.ts
Comment (KvelaGorrrrnio) [inline src/app/shared/services/dialog-confirmation.service.ts]: `switch` feels like an odd word to use here in my opinion. We use `open` for dialogs other places in code.

File: src/app/core/services/basic-rundown-state.service.ts
Comment (KvelaGorrrrnio) [inline src/app/core/services/basic-rundown-state.service.ts]: Instead of having a specific get method on the state service, I would suggest to change to `getBasicRundowns()` and only have it return the basic rundowns array. Then where the `getNonIdleRundown` is used, can do the analysis that it needs. It will make the state service usable by more.

File: src/app/shared/services/dialog-confirmation.service.spec.ts
Comment (KvelaGorrrrnio) [inline src/app/shared/services/dialog-confirmation.service.spec.ts]: To create a test that resembles "real-life" as close as possible, we should not mock value objects like observables. You can use the `EMPTY` observable exposed by RxJS for this purpose.

File: src/app/shared/services/dialog-confirmation.service.spec.ts
Comment (KvelaGorrrrnio) [inline src/app/shared/services/dialog-confirmation.service.spec.ts]: Rundown is misspelled – a `d` is missing :)

File: src/app/shared/services/dialog-confirmation.service.ts
Comment (KvelaGorrrrnio) [inline src/app/shared/services/dialog-confirmation.service.ts]: Instead of duplicating the code with these checks, I suggest that the conditions are combined and simplified. The given rundown has already confirmed that it is not active (on air), so it must be inactive or in rehearsal. Therefore in the second condition, it is enough to check that the ids match, which leaves us with the code:
```typescript
    if (!nonIdleRundown || nonIdleRundown.id === rundown.id) {
      this.dialogService.createConfirmDialog(rundown.name, this.activationMessage, this.activationOkButtonText, () => this.rundownService.activate(rundown.id).subscribe())
      return
    }
```

File: src/app/shared/services/dialog-confirmation.service.ts
Comment (KvelaGorrrrnio) [inline src/app/shared/services/dialog-confirmation.service.ts]: Is there a reason that the color scheme and severity are set here, but not in the other createConfirmDialog calls?

File: src/app/shared/services/dialog-confirmation.service.ts
Comment (KvelaGorrrrnio) [inline src/app/shared/services/dialog-confirmation.service.ts:55]: It is not expressed in the task, but from my understanding we don't won't to be able to go from an active (on air) rundown to rehearsal. Shouldn't that apply for cross-rundown mode transitions as well?
@LindvedKrvang what are your thoughts?

File: src/app/rundown-view/factories/key-binding.factory.ts
Comment (LindvedKrvang) [reply src/app/rundown-view/factories/key-binding.factory.ts]: This method has been changed/deleted in https://github.com/tv2/ng-sofie/pull/142
I highly suggest to not start refactoring it in this PR

File: src/app/shared/services/dialog-confirmation.service.ts
Comment (k-primeholding-dim) [reply src/app/shared/services/dialog-confirmation.service.ts]: When 'deactivate' is involved a 'DANGER' is used.

File: src/app/rundown-view/factories/key-binding.factory.ts
Comment (k-primeholding-dim) [reply src/app/rundown-view/factories/key-binding.factory.ts]: > 
> Sidenote: It also looks like that entering rehearsal have no key binding yet, so I have added it in the suggestion with the key binding `Control + Backquote`.
> 
I was on this same page and introduced (a different 'Shift+Backquote' in resemblance of the MiniShelf cycling in 'opposite' ) key binding and later removed it here https://github.com/tv2/ng-sofie/pull/131/commits/f0338066955a749ebc0e33c5c998ca86751f6fd9 
So at the end this appears to be a really missing feature.

File: src/app/shared/services/dialog-confirmation.service.ts
Comment (LindvedKrvang) [reply src/app/shared/services/dialog-confirmation.service.ts:55]: That's a good question. I can see the argument for why it shouldn't be allowed. But this is a PO question, so we need to ask Jacob.

File: src/app/core/services/basic-rundown-state.service.ts
Comment (LindvedKrvang) [inline src/app/core/services/basic-rundown-state.service.ts:133]: We already have a method in this file that basically does the same as this one: `subscribeToBasicRundowns()`. We should ideally only have one method that gives us the BasicRundowns.

**Review by LindvedKrvang — APPROVED** [review]

File: src/app/core/services/basic-rundown-state.service.ts
Comment (KvelaGorrrrnio) [reply src/app/core/services/basic-rundown-state.service.ts:133]: I think there is a great enough of difference between getting a snapshot of the state and subscribing to the state to have this method.

File: src/app/shared/services/dialog-confirmation.service.ts
Comment (RasmusAlbrektsen) [inline src/app/shared/services/dialog-confirmation.service.ts:26]: If i have a rundown in `REHEARSAL` and i then try to activate it (as "on air"), i will now be asked if that same rundown should be deactivated when i try to do so. I don't know if this is a huge issue, but it is for sure not the way it should be.

File: src/app/shared/services/dialog-confirmation.service.ts
Comment (KvelaGorrrrnio) [reply src/app/shared/services/dialog-confirmation.service.ts:55]: It is actually a product team question, since we don't have a PO ;)

File: src/app/shared/services/dialog-confirmation.service.spec.ts
Comment (KvelaGorrrrnio) [inline src/app/shared/services/dialog-confirmation.service.spec.ts]: Instead of using the magic strings `this` and `another`, they should be extracted to local variables. That way if we change the id of one or both of the rundowns, it won't break the test. It also makes the connection between the mock and the test rundowns clear.

File: src/app/shared/services/dialog-confirmation.service.spec.ts
Comment (KvelaGorrrrnio) [inline src/app/shared/services/dialog-confirmation.service.spec.ts]: `Rundo` should be `Rundown`. This applies to other places in the file as well :)

File: src/app/shared/services/dialog-confirmation.service.spec.ts
Comment (KvelaGorrrrnio) [inline src/app/shared/services/dialog-confirmation.service.spec.ts]: Shouldn't the rundown with id `this` be returned as well?

File: src/app/shared/services/dialog-confirmation.service.spec.ts
Comment (KvelaGorrrrnio) [inline src/app/shared/services/dialog-confirmation.service.spec.ts]: This is not a mock and is an observable, so perhaps this should be renamed to `emptyObservable`.

This should also be used for the other places in the file where the observables are mocked.

File: src/app/shared/services/dialog-confirmation.service.ts
Comment (k-primeholding-dim) [reply src/app/shared/services/dialog-confirmation.service.ts:26]: Great catch.

**Review by RasmusAlbrektsen — APPROVED** [review]

File: src/app/shared/services/dialog-confirmation.service.ts
Comment (k-primeholding-dim) [reply src/app/shared/services/dialog-confirmation.service.ts:55]: So should I wait before I merge in develop, or are we good to go?

File: src/app/shared/services/dialog-confirmation.service.spec.ts
Comment (KvelaGorrrrnio) [inline src/app/shared/services/dialog-confirmation.service.spec.ts:24]: This should be named as the others and use `EMPTY` instead of mocking :)

File: src/app/shared/services/dialog-confirmation.service.spec.ts
Comment (KvelaGorrrrnio) [inline src/app/shared/services/dialog-confirmation.service.spec.ts:53]: This should be named as the others and use `EMPTY` instead of mocking :)

File: src/app/shared/services/dialog-confirmation.service.ts
Comment (KvelaGorrrrnio) [reply src/app/shared/services/dialog-confirmation.service.ts:55]: We are good to go. We'll add it as a feature/bug task if necessary :)

**Review by KvelaGorrrrnio — APPROVED** [review]

**Pull request merged**
